src/dhcp/clientv4.rs

```diff
@@ -0,0 +1,322 @@
+    /// use smoltcp::time::Instant;
+    ///
+    /// let mut sockets = SocketSet::new(vec![]);
+    /// let dhcp_rx_buffer = UdpSocketBuffer::new(
+    ///     [UdpPacketMetadata::EMPTY; 1],
+    ///     vec![0; 1500]
+    /// );
+    /// let dhcp_tx_buffer = UdpSocketBuffer::new(
+    ///     [UdpPacketMetadata::EMPTY; 1],
+    ///     vec![0; 3000]
+    /// );
+    /// let mut dhcp = Dhcpv4Client::new(
+    ///     &mut sockets,
+    ///     dhcp_rx_buffer, dhcp_tx_buffer,
+    ///     Instant::now()
+    /// );
+    /// ```
+    pub fn new<'a, 'b, 'c>(sockets: &mut SocketSet<'a, 'b, 'c>, rx_buffer: UdpSocketBuffer<'b, 'c>, tx_buffer: UdpSocketBuffer<'b, 'c>, now: Instant) -> Result<Self>
+    where 'b: 'c,
+    {
+        let mut udp_socket = UdpSocket::new(rx_buffer, tx_buffer);
+        udp_socket.bind(IpEndpoint { addr: IpAddress::Unspecified, port: UDP_CLIENT_PORT })?;
+        let udp_handle = sockets.add(udp_socket);
+
+        Ok(Client {
+            state: ClientState::Discovering,
+            udp_handle,
+            next_egress: now,
+            transaction_id: 1,
+        })
+    }
+
+    /// When to send next packet
+    ///
+    /// Useful for suspending execution after polling.
+    pub fn next_poll(&self, now: Instant) -> Duration {
+        self.next_egress - now
+    }
+
+    /// Process incoming packets on the contained UdpSocket, and send
+    /// DHCP requests when timeouts are ready.
+    ///
+    /// Applying the obtained network configuration is left to the
+    /// user. You must configure the new IPv4 address from the
+    /// returned `Config`. Otherwise, DHCP will not work.
+    ///
+    /// A Config can be returned from any valid DHCP reply. The client
+    /// performs no bookkeeping on configuration or their changes.
+    pub fn poll<DeviceT: for<'d> Device<'d>>
+               (&mut self,
+                iface: &mut Interface<DeviceT>, sockets: &mut SocketSet,
+                now: Instant) -> Result<Option<Config>> {
+        let mut udp_socket = sockets.get::<UdpSocket>(self.udp_handle);
+
+        // Process incoming
+        let config = if udp_socket.can_recv() {
+            let (data, src) = udp_socket.recv()?;
+            self.ingress(iface, now, data, &src)
+        } else {
+            None
+        };
+
+        // Send requests
+        if udp_socket.can_send() && now >= self.next_egress {
+            self.egress(iface, &mut *udp_socket, now)?;
+        }
+
+        Ok(config)
+    }
+
+    fn ingress<DeviceT: for<'d> Device<'d>>
+              (&mut self,
+               iface: &mut Interface<DeviceT>, now: Instant,
+               data: &[u8], src: &IpEndpoint) -> Option<Config> {
+        let src_ip = match src.addr {
+            IpAddress::Ipv4(src_addr) => src_addr,
+            _ => return None,
+        };
+
+        let dhcp_packet = match DhcpPacket::new_checked(data) {
+            Ok(dhcp_packet) => dhcp_packet,
+            Err(_) => return None,
```

Comment:
Elsewhere, invalid replies are being logged, which helps with debugging. Can you please add such code here as well?